openssh: add UCI support

mhusaam opened the pull request:
Update init script to handle UCI and add a default config

Signed-off-by: Mohd Husaam Mehdi <husaam.mehdi@iopsys.eu>

mhusaam committed:
openssh: add UCI support

* update init script to handle sshd UCI by generating config file
  and passing that config file to the sshd daemon
* add a default sshd config

Signed-off-by: Mohd Husaam Mehdi <husaam.mehdi@iopsys.eu>

stokito commented on net/openssh/files/sshd.init:
```diff
@@ -20,19 +111,28 @@ start_service() {
 	mkdir -m 0700 -p /var/empty
+	mkdir -m 0700 -p /root/.ssh
 
-	local lport=$(awk '/^Port / { print $2; exit }' /etc/ssh/sshd_config)
```
This is needed to read the Port from sshd_config for mdns.
The Port from UCI may be overridden by the sshd_config.
Not so critical but keep in mind.

stokito replied:
But generally the problem is that old sshd_config may be overridden by the options from the default config that was installed with a new version of a package.

mhusaam replied:
Okay, thank you.

stokito commented on net/openssh/files/sshd.init:
```diff
@@ -6,8 +6,99 @@ STOP=50
+	[ "${mdns}" -ne 0 ] && procd_add_mdns "ssh" "tcp" "${Port}" "daemon=sshd"
+
+	if [ "${IdleTimeout}" -ne 0 ]; then
+		procd_append_param command -o "ClientAliveCountMax 1"
```
`ClientAliveCountMax 1` may be too small.
By default it's 3 so maybe keep it as is?

https://manpages.debian.org/stretch/openssh-server/sshd_config.5.en.html#ClientAliveCountMax

mhusaam replied:
This is my logic:

from https://www.man7.org/linux/man-pages/man5/sshd_config.5.html

**ClientAliveCountMax:**
The default value is 3.  If ClientAliveInterval is set to
15, and ClientAliveCountMax is left at the default,
unresponsive SSH clients will be disconnected after
approximately 45 seconds.  Setting a zero
ClientAliveCountMax disables connection termination.

Therefore, to mimic IdleTimeout from dropbear, we set
ClientAliveCountMax to 1 and ClientAliveInterval to
the time we want to wait.

stokito commented on net/openssh/files/sshd.init:
```diff
@@ -6,8 +6,99 @@ STOP=50
+		'Port:port:22' \
+		'IdleTimeout:uinteger:0' \
+		'MaxAuthTries:uinteger:3' \
+		'enable:bool:1' \
```
`enable` -> `enabled`

mhusaam replied:
It was enable in dropbear so I though I should keep it enable here too.

stokito commented on net/openssh/files/sshd.config:
```diff
@@ -0,0 +1,6 @@
+config sshd
+	option enable '1'
```
`enable` -> `enabled`

mhusaam replied:
It was enable in dropbear so I thought I should keep it enable here too.

stokito commented:
The PR should be useful given that vendors of modern routers prefer the OpenSSH.

Currently we already have the `/etc/config/dropbear` config that has similar keys:
https://github.com/openwrt/openwrt/blob/main/package/network/services/dropbear/files/dropbear.init#L165

Maybe we can migrate the dropbear config to the sshd. E.g. first check if the dropbear exists and if not read the sshd. And similarly for the OpenSSH but it will install own config and that will cause a conflict which config to use.

Users may change the ssh server e.g. have the Dropbear and then install the OpenSSH. But old settings won't be used now e.g. Port number. Of course a user should know what he doing but still some people may lost an access.

Anyway, that would be extremely useful to support same options that the dropbear supports currently. Well, maybe not all, like the `BannerFile` is not critical at all.

Basically since the Dropbear imitates the sshd. It should be possible to just take its init file and just replace PROG and it must work. If something is not compatible we may patch the Dropbear as I did recently for the ssh-keygen.

mhusaam commented:
> The PR should be useful given that vendors of modern routers prefer the OpenSSH.
> 
> Currently we already have the `/etc/config/dropbear` config that has similar keys: https://github.com/openwrt/openwrt/blob/main/package/network/services/dropbear/files/dropbear.init#L165
> 
> Maybe we can migrate the dropbear config to the sshd. E.g. first check if the dropbear exists and if not read the sshd. And similarly for the OpenSSH but it will install own config and that will cause a conflict which config to use.
> 
> Users may change the ssh server e.g. have the Dropbear and then install the OpenSSH. But old settings won't be used now e.g. Port number. Of course a user should know what he doing but still some people may lost an access.
> 
> Anyway, that would be extremely useful to support same options that the dropbear supports currently. Well, maybe not all, like the `BannerFile` is not critical at all.
> 
> Basically since the Dropbear imitates the sshd. It should be possible to just take its init file and just replace PROG and it must work. If something is not compatible we may patch the Dropbear as I did recently for the ssh-keygen.

Hello, thank you for reviewing this PR. I am sorry for the delayed response.

I have updated the PR as follows:
- I have added support for migrating from dropbear to sshd. So, if someone has an image with dropbear and they upgrade their image to one which has sshd, sshd will look for dropbear config file, and if it is found, it will try to generate sshd config from that, and also try to copy the authorized_keys file.
- I have added support for more dropbear options.
- I have changed the approach from passing command line arguments to sshd daemon, to generating a sshd config file and passing that config file.

Kindly review.

stokito commented:
Oh, sorry, I didn't noticed that the `IdleTimeout` is a Dropbear's thing https://www.mankier.com/8/dropbear#-I

One thing that I see here that if in a future we'll add the `ClientAliveCountMax` then we may have a conflict and add some priority for them. Also this may confuse that the option from the Dropbear is used for OpenSSH.
What if we simply rename it to `ClientAliveInterval` during migration? Yes, technically this will be not exactly the same and will effectively make the timeout triple bigger but that should not break anything. But code will be much simpler and smaller.
 
The LUCI doesn't show an option for the `IdleTimeout` and it's also not in the default Dropbraer's config. Not sure how many users use it. So maybe it's not that important.

stokito commented on net/openssh/files/sshd.init:
```diff
@@ -6,8 +6,351 @@ STOP=50
+	}
+
+	local PidFile="/var/run/${NAME}.${Cfg}.pid"
+	local ConfigFile="/etc/ssh/${NAME}.${Cfg}.conf"
```
The /etc/ folder is persistent and (just potentially) we may kill NAND flash with writes on each restart.
Maybe instead we can put the generated config into `/var/run/sshd` e.g. into tmpfs.

mhusaam replied:
Thanks for pointing this out, I have updated, kindly review.

stokito commented:
I grepped the `option enable `, `config_get enable ` and `config_get_bool enable ` in packages and we do have a few configs that are using the `enable` instead of `enabled`. So to avoid accidental SSH shell loss after editing of the conf I think it would be better to check **both** `enable` and `enabled`.
In a documentation always use the `enabled`. I also think it would be good to make same for other places that use the `enable`: `atftpd`, `collectd`, `dnsmasq`, `dropbear`, `natmap`, `nut`, `socat`, `sslh`, `openvpn`.
Please let me know what you think about this.

For bool options must be used `config_get_bool` instead of just `config_get`. It will parse `no`/`off` `yes`/`on`.

stokito commented on net/openssh/files/sshd.init:
```diff
@@ -6,8 +6,351 @@ STOP=50
+load_interfaces()
+{
+	config_get Interface "$1" Interface
+	config_get enable "$1" enable 1
```
config_get_bool enable "$1" enable 1

mhusaam replied:
Hello, kindly review:

- added a check for both enable and enabled
- updated to use config_get_bool instead of config_get

stokito commented:
I checked and the Dropbear also has some options that weren't migrated:
* `ForceCommand`
* `rsakeyfile` but it was deprecated
* `keyfile` corresponds to `HostKeyFiles`
* `SSHKeepAlive` corresponds to (again) the `ClientAliveInterval`
* `RecvWindowSize` doesn't have a similar option in sshd

mhusaam commented:
> Oh, sorry, I didn't noticed that the `IdleTimeout` is a Dropbear's thing https://www.mankier.com/8/dropbear#-I
> 
> One thing that I see here that if in a future we'll add the `ClientAliveCountMax` then we may have a conflict and add some priority for them. Also this may confuse that the option from the Dropbear is used for OpenSSH. What if we simply rename it to `ClientAliveInterval` during migration? Yes, technically this will be not exactly the same and will effectively make the timeout triple bigger but that should not break anything. But code will be much simpler and smaller.
> 
> The LUCI doesn't show an option for the `IdleTimeout` and it's also not in the default Dropbraer's config. Not sure how many users use it. So maybe it's not that important.

Hello,

- I think what is necessary is that the meaning of the option is preserved. 
- Also, like you said, IdleTimeout is only for dropbear and ClientAliveCountMax and ClientAliveInterval are only for sshd (OpenSSH).
- So either the user can use dropbear option or the explicit sshd options.
- Accordingly I have **updated the logic to give higher priority to sshd options if they are set explicitly.**

mhusaam commented:
> I checked and the Dropbear also has some options that weren't migrated:
> 
> * `ForceCommand`
> * `rsakeyfile` but it was deprecated
> * `keyfile` corresponds to `HostKeyFiles`
> * `SSHKeepAlive` corresponds to (again) the `ClientAliveInterval`
> * `RecvWindowSize` doesn't have a similar option in sshd
> 
> This is not critical at all, the PR can be merged and we can back to thos later.

I have added support for the option **keyfile** in init script.
But I did not add it to migration script because uci_validate checks the presence of the file and if the file is not present then init script will not start sshd.
So, if the option is set and somehow the file is removed during upgrade (maybe because it was present in non-persistent storage, I know the chance of that is low) then ssh will not be possible to the device unless defaultreset is done.

As you said, maybe we can come back to this later.

mhusaam commented:
> When installing the sshd it will copy options from dropbear and remove (!) the dropbear config. But the dropbear will remain started and the port 22 will be busy.
> 
> I think that on migration we should copy options from the dropbear but disable the sshd e.g. `uci set sshd.main.enabled=0`. Then a user can itself go and disable dropbear and enable the sshd and make the `uci commit` and restart the sshd. At least he will be in control and know what he is doing.
> 
> Also it looks like we need a protection from double migration after OpenWrt firmware upgade:
> 
> https://openwrt.org/docs/guide-developer/uci-defaults#ensuring_scripts_don_t_overwrite_custom_settingsimplementing_checks
> 
> The least but not least. What if a user just manualy configured the `/etc/ssh/sshd_config` and then upgreade the `openssh-server` package and now instead of his config the `/etc/config/sshd` is used.
> 
> We may need to add an option `config_path /etc/ssh/sshd_config` that will translated to `Include /etc/ssh/sshd_config`.

Hello,

Sorry for the delayed response.

We need to consider mirgation on upgrade too.

If we disable sshd, and someone upgrades with an image which has sshd (openssh), then sshd will get disabled and there will be no ssh possible to the device.

I have added support for option IncludeConfigFile.

Kindly review.

feckert commented:
- please rebase the changes
- stash to one commit
- please bump PKG_RELEASE number by one

mhusaam commented:
Updated, kindly review.

GeorgeSapkin commented:
@mhusaam still relevant? There are merge conflicts.

GeorgeSapkin commented on net/openssh/Makefile:
```diff
@@ -244,6 +244,8 @@ define Package/openssh-server/install
 	$(INSTALL_DIR) $(1)/etc/init.d
 	$(INSTALL_BIN) ./files/sshd.init $(1)/etc/init.d/sshd
+	$(INSTALL_DIR) $(1)/etc/config
+	$(INSTALL_BIN) ./files/sshd.config $(1)/etc/config/sshd
```
```suggestion
	$(INSTALL_CONF) ./files/sshd.config $(1)/etc/config/sshd
```

mhusaam replied:
Done.

GeorgeSapkin commented on net/openssh/files/sshd.init:
```diff
@@ -6,8 +6,358 @@ STOP=50
+{
+	local IpAddrs
+	local Cfg="$1"
+	local ValidationResult="${2}"
```
It looks odd when you mix different variable uses, e.g. `$1` and `${2}`. Pick one style and stick to it, unless it's strictly necessary to use the other, which is not the case here.

mhusaam replied:
Updated, kindly review.

GeorgeSapkin reviewed (changes requested):
On top of all the other small comments, please run a shellcheck on all the scripts, if you haven't already.

systemcrash commented on net/openssh/files/sshd.init:
```diff
@@ -6,8 +6,358 @@ STOP=50
+		'Port:port:22' \
+		'RootLogin:bool:1' \
+		'RootPasswordAuth:bool:1' \
+		'UsePAM:bool:0'
```
They just look different because they use capitalisation, which is not the common parlance in uci.

mhusaam replied:
These mostly refer to sshd config manual. I think it is too open ended to specify all possible config values in this case.

GeorgeSapkin replied:
Perhaps add a note of that then?

mhusaam commented:
Ran shellcheck, it only found benign stuff like "local is not supported".

pprindeville commented on net/openssh/files/sshd.config:
```diff
@@ -0,0 +1,83 @@
+# Below are UCI usage examples and defaults.
+
+# AllowUsers: list of strings (default: empty)
+# list AllowUsers 'root'
```
If these are subordinate to `config sshd` then please indent them.

pprindeville commented on net/openssh/files/sshd.init:
```diff
@@ -27,19 +356,32 @@ start_service() {
 		}
 	done
 	mkdir -m 0700 -p /var/empty
+	mkdir -m 0700 -p /root/.ssh
```
Should this go here or in `base-files`?

pprindeville reviewed (commented):
Where are we placing the synthetic files?  Into `/etc/ssh/sshd_config.d/*` or somewhere else?
